### ISSUE TITLE
music visualizer

### ISSUE THREAD

**Issue by alltagshack** [Starcommander/StreamRadio#7]
Hi.

I did it again and build a new "sweded" [version](https://github.com/no-go/StreamRadio). I import some files from https://github.com/GautamChibde/android-audio-visualizer and got after small modifications a music visualisation. The "android-audio-visualizer" project is not GNU license, but Apache 2. Just adopting some file sould not run in any trouble (?)

- maybe a new feature for the future
- maybe reduce battery to fast
- need one additional permission :-(

I am sorry, but my master thesis is on my agenda again and I will leave you alone with your WebRadio App. I hope, I could support you a little bit with some tweaks ;-)

Viele Grüße vom Niederrhein
Jochen
![photo](https://user-images.githubusercontent.com/12084008/55269401-e0b26000-5292-11e9-83ec-203c685f015b.png)

**Comment by Starcommander** [comment]
It has nothing to do with "Just adopting some file" , because each line of code is licensed in apache2.
Basically it should be possible to include this library without any troubles.
http://www.apache.org/licenses/GPL-compatibility.html

Maybe you can supply the differences for this nice changes?!
Next time I will try to create a new Version for FDroid.
Thanks for your very helpful work.

**Comment by alltagshack** [comment]
> Thanks for your very helpful work.

:-D

ok, I try to explain my copy/paste decisions a bit:

First I search a opensource app on fdroid to look into the "equilzer" code, because another issue describes the system-wide equilizer as 'it is easy' (?). I found https://github.com/iamSahdeep/Bop with a "to complex" equlizer and decide to take a look into the "visualizer" code. Surprise! There is a lib for that! ... but after hours of work and try to fix: the lib eplainations does not work!!!

hm. What happend? I look to the [expaination](https://github.com/GautamChibde/android-audio-visualizer#importing-the-library) how to add to the dependencies and the [sample code description](https://github.com/GautamChibde/android-audio-visualizer/wiki/Bar-Visualizer). After doing this everything seams to work. But: App crashed very early with some "DEX zip ... BarViritualizer not found". After a long time to fix it, I decided to look into the "bob app" code: Ups!

- bob app: `audiovisualizer` not in dependencies section
- bob app: [visualizer](https://github.com/iamSahdeep/Bop/blob/master/app/src/main/java/com/sahdeepsingh/Bop/Visualizers/visualizer.java) code looks like the `audiovisualizer` lib code
- audiovisualizer sample app: `audiovisualizer` not in dependencies section

Because of the 2. point and the only magic in the code seams to come from `android.media.audiofx.Visualizer` I decide to include the lib code into my app by drag-and-dropping it from the zip-download into my intellij IDE:

- [original BaseVisualizer](https://github.com/GautamChibde/android-audio-visualizer/blob/master/audiovisualizer/src/main/java/com/chibde/BaseVisualizer.java) to [my BaseVisualizer](https://github.com/no-go/StreamRadio/0a9dc83c4f8b87a1a977698be2f5adc429948356/master/app/src/main/java/starcom/snd/sweded/BaseVisualizer.java) ... but I remove the "Nullable" stuff/hint
- [original visualizer/BarVisualizer](https://github.com/GautamChibde/android-audio-visualizer/blob/master/audiovisualizer/src/main/java/com/chibde/visualizer/BarVisualizer.java) to [my visualizer/BarVisualizer](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/src/main/java/starcom/snd/sweded/visualizer/BarVisualizer.java) ... but I remove the "Nullable" stuff/hint, too

Because of the inline code hints and inline license hints I had no hesitation in copying that.

After doing that:

- I just add these new type of view into the [layout](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/src/main/res/layout/activity_main.xml#L54)
- I access and configure that view in the [Activity](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/src/main/java/starcom/snd/sweded/WebRadio.java#L83) 
- I add the permission the listen to sound [here](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/src/main/AndroidManifest.xml#L24)
- I add [code](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/src/main/java/starcom/snd/sweded/WebRadio.java#L182) to bind the Visualizer to the PlayerSession only once! (AND only on State=Playing)
- finally to get the session from your `WebStreamPlayer` I made the method [getMediaPlayer](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/src/main/java/starcom/snd/sweded/WebStreamPlayer.java#L67) public

The only bug I found: if the sound plays and the visualizer work, switching the dark/day theme crashes the App

I hope, I have not forgotten anything. maybe you have use the new 28 [appcombat](https://github.com/no-go/StreamRadio/blob/0a9dc83c4f8b87a1a977698be2f5adc429948356/app/build.gradle#L30)

**Comment by alltagshack** [comment]
> The only bug I found: if the sound plays and the visualizer work, switching the dark/day theme crashes the App

The trick to fix this bug: doing release for recreate() in the 2 action items:

```java
 ...
 item.setChecked(false);
 mPreferences.edit().putBoolean("is_dark", false).apply();
 getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO);
 barVisualizer.release();
 recreate();
```

**Comment by alltagshack** [comment]
Hi. A friend test a modified App and got an "App crash issue" on changing to night/day mode.

I only fix this by a real "app close and reopen":

```java
...
 item.setChecked(false);
 mPreferences.edit().putBoolean("is_dark", false).commit();
 getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO);
 Intent mStartActivity = new Intent(this, WebRadio.class);
 int mPendingIntentId = 123456;
 PendingIntent mPendingIntent = PendingIntent.getActivity(this, mPendingIntentId, mStartActivity, PendingIntent.FLAG_CANCEL_CURRENT);
 AlarmManager mgr = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
 mgr.set(AlarmManager.RTC, System.currentTimeMillis() + 500, mPendingIntent);
 System.exit(0);
```

Not a good solution :-S and the +500ms feels like 3sec

**Comment by Starcommander** [comment]
No, the solution is not good!
To find the reason of app crash we need a log-output.
With android-sdk you can connect the phone, and enter "adb logcat" to get the logs.
Then you should be able to find an exception or a stack trace for this issue.
(look at the timestamps)

(An other way is to just drop the visualizer that needs accu-power and additional permission)

**Comment by alltagshack** [comment]
I know about the logging, but without the device which result in that crash issue, it is impossible :-S

I did not even know, if it is the result of the visualizer or a "recreate()" issue special on android 9.
